### PR TITLE
Listen to webglcontextcreationerror event and give detailed debug info when it fails

### PR DESCRIPTION
**Pull request by zhangyiatmicrosoft**
When we have the control hosted and deployed, there could be many end users experiencing “Failed to initialize WebGL” error for various reasons. We cannot debug their machine and the current error is too generic. 
Adding this listener is a great help and it does not affect performance, as far as I can tell.

 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality.
 - [x] Post benchmark scores.
 - [x] Manually test the debug page.
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Comment by zhangyiatmicrosoft** [comment]
main 7db0e81                 main 3b102b7  addwebgldebug-dirty e603e0c
                       Paint                  9.1303 ms                    9.8256 ms⚠️                   9.6477 ms⚠️
                  QueryPoint                 15.4423 ms⚠️                  15.1823 ms⚠️                  16.3600 ms⚠️
                    QueryBox                137.3285 ms                  137.0523 ms                  140.7258 ms
                      Layout                150.9712 ms                  137.2806 ms                  138.5921 ms
                   Placement                  9.4746 ms                    9.3711 ms⚠️                   9.3924 ms
                    Validate                  2.0450 ms                    2.0092 ms                    2.0628 ms⚠️
            StyleLayerCreate                  0.6018 ms                    0.7341 ms⚠️                   0.7532 ms
              FunctionCreate                  0.0247 ms                    0.0249 ms                    0.0281 ms⚠️
            FunctionEvaluate                  0.0032 ms                    0.0031 ms                    0.0032 ms
            ExpressionCreate                  0.3461 ms                    0.3483 ms                    0.4371 ms⚠️
          ExpressionEvaluate                  0.0049 ms                    0.0049 ms                    0.0065 ms☠️
              WorkerTransfer                 11.9881 ms⚠️                  91.4519 ms                   92.3015 ms
                 PaintStates                 67.6664 ms                   67.3770 ms                   67.3945 ms
         PropertyLevelRemove                 66.7000 ms                   66.9085 ms                   67.3378 ms
          FeatureLevelRemove                 66.9628 ms                   66.4724 ms                   66.5008 ms
           SourceLevelRemove                 66.7195 ms                   67.4519 ms                   68.0500 ms
             LayerBackground                  1.6913 ms                    3.7127 ms⚠️                   1.7009 ms
                 LayerCircle                  0.1664 ms                    0.1708 ms                    0.1730 ms
                   LayerFill                  0.3216 ms⚠️                   0.3123 ms                    0.3101 ms
          LayerFillExtrusion                  0.5316 ms                    0.5328 ms                    0.5343 ms
                LayerHeatmap                  0.3754 ms                    0.3765 ms⚠️                   0.3768 ms
              LayerHillshade                  0.5224 ms                    0.5228 ms                    0.5276 ms
                   LayerLine                  0.1853 ms                    0.1901 ms                    0.1873 ms
                 LayerRaster                  2.1212 ms                    2.1637 ms                    2.1603 ms
                 LayerSymbol                  0.6704 ms                    0.6545 ms                    0.6678 ms
        LayerSymbolWithIcons                  0.7311 ms                    0.7256 ms                    0.7501 ms
 LayerTextWithVariableAnchor                 14.2692 ms                   14.2382 ms                   14.1409 ms
      LayerSymbolWithSortKey                  0.3960 ms☠️                   0.3944 ms☠️                   0.3913 ms☠️
                        Load                 31.7000 ms                   32.3859 ms                   32.2283 ms
                SymbolLayout                  7.8669 ms                    7.8674 ms                    7.8312 ms
                FilterCreate                  0.4631 ms                    0.4561 ms                    0.4622 ms
              FilterEvaluate                  0.3902 ms                    0.4154 ms                    0.4261 ms
               HillshadeLoad                348.3141 ms                  345.4912 ms                  345.9824 ms
                 CustomLayer                                               0.0217 ms                    0.0224 ms
                     MapIdle                                             665.3441 ms                  696.6441 ms

**Comment by zhangyiatmicrosoft** [comment]
Full benchmark results (all.pdf) is big, do people usually upload it?
What is the best way to post text only version?

Dont know why GitHub is formatting first half of the benchmark results but not the rest. Here they are:
image

File: src/ui/map.ts [hunk 1]
```diff
@@ -2545,24 +2545,35 @@ class Map extends Camera {
...
+                errorDetails.type = args.type;
+            }
+            webglcontextcreationerrorDetails = JSON.stringify(errorDetails);
+        }, {once: true});
+
         const gl = this._canvas.getContext('webgl', attributes) ||
             this._canvas.getContext('experimental-webgl', attributes);
 
         if (!gl) {
-            this.fire(new ErrorEvent(new Error('Failed to initialize WebGL')));
```
Comment (zhangyiatmicrosoft) [inline src/ui/map.ts:2559]: "this.fire" just calls console.error for error events. In this case it is duplicated so deleted.

File: src/ui/map.ts [hunk 2]
```diff
@@ -2545,24 +2545,35 @@ class Map extends Camera {
         this._canvas.style.height = `${height}px`;
     }
 
-    _setupPainter() {
+    _setupPainter(): string {
```
Comment (HarelM) [inline src/ui/map.ts:2548]: Why not simply throw inside this method instead of returning the string?

Comment (zhangyiatmicrosoft) [reply src/ui/map.ts]: Good idea

> Why not simply throw inside this method instead of returning the string?

File: src/ui/map.ts [hunk 3]
```diff
@@ -2552,11 +2549,19 @@ class Map extends Camera {
             antialias: this._antialias || false
         });
 
+        this._canvas.addEventListener('webglcontextcreationerror', (args: WebGLContextEvent) => {
+            const errorDetails: any = {requestedAttributes: attributes};
+            if (args) {
+                errorDetails.statusMessage = args.statusMessage;
+                errorDetails.type = args.type;
+            }
+            throw new Error(JSON.stringify(errorDetails));
```
Comment (HarelM) [inline src/ui/map.ts:2558]: This looks like a throw in a middle of a callback, will it throw in the original `_setupPainter` method?
I would recommend to try and reproduce this somehow or at least simulate this in a unit test. Adding a unit test will also make sure that this behavior is kept if someone else changes this in the future...

**Comment by HarelM** [comment]
Also a changelog entry is needed.
note that a PR has a template for a reason...

**Comment by wipfli** [comment]
Maybe I should revisit the pull request templeate. @zhangyiatmicrosoft do you have any feedback on the template?

**Comment by HarelM** [comment]
@wipfli while you do so, please change the instructions related to changelog file, they are cryptic at best right now... 😀

File: src/ui/map.ts (hunk 3 above)
Comment (zhangyiatmicrosoft) [reply src/ui/map.ts]: Thanks for the suggestion. Fixed and added UT

**Comment by zhangyiatmicrosoft** [comment]
> Also a changelog entry is needed. note that a PR has a template for a reason...

Did not know exactly which change should be included in changelog. :)
Fixed. Thank you

**Comment by zhangyiatmicrosoft** [comment]
> Maybe I should revisit the pull request templeate. @zhangyiatmicrosoft do you have any feedback on the template?

This item is not clear to me:
Add an entry inside this element for inclusion in the maplibre-gl-js changelog: \<changelog\>\</changelog\>

Am I supposed to update changelog.md? Or add some content inline and other process will pick it up?
Checked a few existing PRs and feels like should update changelog.md manually, so I did that. Please let me know what the correct way is. Thanks

File: CHANGELOG.md [hunk 4]
```diff
@@ -12,6 +12,7 @@
 - Return a promise from `once` method to allow easier usage of async/await in this case ([#1690(https://github.com/maplibre/maplibre-gl-js/pull/1690))
 - Add pseudo (CSS) fullscreen as a fallback for iphones ([#1678](https://github.com/maplibre/maplibre-gl-js/pull/1678))
 - Add `updateData` to `GeoJSONSource` which allows for partial data updates ([#1605](https://github.com/maplibre/maplibre-gl-js/pull/1605))
+- Listen to webglcontextcreationerror event and give detailed debug info when it fails ([#1715](https://github.com/maplibre/maplibre-gl-js/pull/1715))
```
Comment (wipfli) [inline CHANGELOG.md:15]: Sorry for nagging, but the `3.0.0-pre.1` section is for the already released version. Please move your entry up to the `### main` section...

Comment (wipfli) [reply CHANGELOG.md:15]: Thanks!

**Comment by wipfli** [comment]
#1729

**Comment by zhangyiatmicrosoft** [comment]
@HarelM please take a look again? Thanks

**Review by HarelM — APPROVED** [review]
Looks great now, thanks!

**Comment by HarelM** [comment]
@JannikGM it's hard to understand what you are suggesting, please open a PR or an issue with the relevant changes...

**Pull request merged**